fix(combobox): async options highlight

J-Michalek opened the pull request:
Fix: https://github.com/unovue/reka-ui/issues/1853

To my understanding the issue here is that the first item is highlighted right away as a reaction to the input event inside of `ListboxFilter` but if the options are only loaded after the input has been set the first option won't get the highlight. To remedy this I used the `highlightFirstItem` function inside of `ComboboxRoot `which is invoked after the search value or number of items changes.

As a sidenote I removed the event parameter from the `highlightFirstItem` function as it does not use it at all but it was required in the types.

J-Michalek committed:
refactor(listbox-root): remove unused parameter from highlightFirstItem function

The function does not use the event in any case.

J-Michalek committed:
fix(combobox-root): highlight first items after item amount is changes

This should fix issues with item highlight while using async loading for options.

J-Michalek committed:
refactor(listbox-virtualizer): fix type-check

J-Michalek commented:
In hindsight I think the usage of the `highlightFirstItem` function should be moved out of `ComboboxRoot` to the `ComboboxInput` as I think this feature is not really wanted when there is no `ComboboxInput` present.

jesseleite commented:
Hey @J-Michalek, thank you for jumping on this! ❤️‍🔥

I'm just playing with your PR here...

Combining my initial StackBlitz example in #1853 with the changes in this PR in package.json, it seems it is greatly improved. I was able to make it fail, but super rarely. For example, I'm not using arrow keys at all here, and it's almost always auto-selecting just fine...

![CleanShot 2025-04-28 at 11 39 41](https://github.com/user-attachments/assets/5467938f-fb69-4854-9127-4b45202dcc6a)

However, if increase the timeout to 2000ms, even with a more proper debounce function (here's a [StackBlitz example](https://stackblitz.com/edit/9c4y9ocg-7bete5jx?file=src%2FApp.vue) for you), I can make it glitch out pretty easily by changing my search query. I realize 2000ms is overkill, but it seems to highlight the issue more clearly...

![CleanShot 2025-04-28 at 11 46 02](https://github.com/user-attachments/assets/ea71d667-d4b6-4409-b633-7121ae308ed5)

Weirdly, it's also rendering `<ComboboxEmpty>`'s "no options" along with the successfully queried option result 🤔

Curious on your thoughts?

J-Michalek commented:
@jesseleite Yeah this is bad. Seems like the code has several race conditions (since the `ComboboxEmpty` gets stuck there) and probably needs a closer look into the internals. This is definitely an issue with using watchers and/or syncing several values resulting in unpredictable state when one thing changes out of the ordinary - in this case the options.

I will try to look more closely into this but no promises as I think I would need to know more about the `Combobox`'s desired functionality to not break other cases which are expected.

Edit: I looked into this a little more and investigated how NuxtUI uses it as their `SelectMenu` component seems to work as expected with async options and I found out that if you use the `ignoreFilter` prop on `ComboboxRoot` then it starts to work as expected. Seems like the issue is somewhere in the internal filtering logic of the `Combobox`.

jesseleite commented:
> I found out that if you use the `ignoreFilter` prop on `ComboboxRoot` then it starts to work as expected. Seems like the issue is somewhere in the internal filtering logic of the `Combobox`.

@J-Michalek We're actually using the ignoreFilter prop ourselves, because we're using [fuzzysort](https://github.com/farzher/fuzzysort) as our search/sort algorithm with Combobox. Honestly, it works great, aside from the issue in #1853. I intentionally left ignoreFilter/fuzzysort out of the StackBlitz implementation, because it didn't seem to matter, and I was trying to keep the reproducible example as simple as possible.

I'm curious what you're doing differently that ignoreFilter works for you as expected?

J-Michalek commented:
@jesseleite Oh I see now, it worked for me because I was using the second example you mentioned [here](https://github.com/unovue/reka-ui/pull/1854#issuecomment-2835718817) which includes the content of this PR and there it seems to work when used together with `ignoreFilter` prop, but without this PR it is broken.

jesseleite commented:
@J-Michalek Ah whoops I'm sorry, yes that example should've had `ignoreFilter`. I've added that prop to [the StackBlitz example](https://stackblitz.com/edit/9c4y9ocg-7bete5jx?file=src%2FApp.vue), and it definitely resolves the `ComboboxEmpty` issue I mentioned above, but it still has trouble highlighting. For example...

- Type 'apple' quickly and wait for highlighted result
- Hit cmd-backspace to clear input text
- Type 'honey' quickly, and it doesn't highlight first result 🤔

![CleanShot 2025-04-29 at 12 09 47](https://github.com/user-attachments/assets/fe3be65b-0961-42a6-8743-2bf291c07cc6)

jesseleite commented:
@J-Michalek Ah whoops I'm sorry, yes that example should've had `ignoreFilter`. I've added that prop to [the StackBlitz example](https://stackblitz.com/edit/9c4y9ocg-7bete5jx?file=src%2FApp.vue), and it definitely resolves the `ComboboxEmpty` issue I mentioned above, but it still has trouble highlighting. For example...

- Type 'apple' quickly and wait for highlighted result
- Hit cmd-backspace to clear input text
- Type 'honey' quickly, and it doesn't highlight first result 🤔

![CleanShot 2025-04-29 at 12 09 47](https://github.com/user-attachments/assets/fe3be65b-0961-42a6-8743-2bf291c07cc6)

I'm also using your latest PR commit in my StackBlitz's package.json, copied from the [bot post above](https://github.com/unovue/reka-ui/pull/1854#issuecomment-2832150713)...

```
"reka-ui": "https://pkg.pr.new/unovue/reka-ui/reka-ui@966dfeb"
```

J-Michalek commented:
@jesseleite Yes, I just wanted to clear any possible confusion. I already got something cooking in terms of refactoring and fixes. I will try to push it most likely tomorrow if I am satisfied with the results.

Commit:
refactor(combobox): use computed instead of reactive with watchers

Using a computed intead of a reactive with updates via watchers
should be more reliable and less prone to issues with stale state.

This also seems to fix the issues we were discussing in this PR.

J-Michalek commented:
@jesseleite I actually got the changes going today, I would very much appreciate your help with testing them. I was not able to break it with the use cases you have described.

The code will need more review but for quick assesment of improvement it should be enough.

jesseleite commented:
Oh nice! Looks like there's conflicts on this branch. Also, is there a way to kick [pkg.pr.new](https://github.com/unovue/reka-ui/pull/1854#issuecomment-2832150713) CI to build for your latest commit? I'm getting this in StackBlitz...

![CleanShot 2025-04-29 at 15 54 47](https://github.com/user-attachments/assets/71666874-244f-441d-ac65-ab9066934ca7)

J-Michalek committed:
fix(combobox-cancel): use filterSearch instead of filterState.search

This happened in a badly resolved merge conflict.

J-Michalek commented:
@jesseleite Hello, I think every change should be released but due to the fact that there were merge conflicts the pipeline did not run. I resolved the conflicts and the changes should be available with https://pkg.pr.new/reka-ui@f161dbd (npm i https://pkg.pr.new/reka-ui@f161dbd).

jesseleite commented:
@J-Michalek Testing with latest commit in this updated [StackBlitz](https://stackblitz.com/edit/9c4y9ocg-ook9vzgu?file=src%2FApp.vue), and I can't get it to fail... Working great! ❤️‍🔥👏

J-Michalek committed:
test(combobox): add story for async options

J-Michalek commented:
I will need to update the PR description to reflect what was actually changed as it is now outdated.

J-Michalek commented:
I decided not to update the PR description so that the conversaion here can be tracked correctly.

@zernonia I would love a review if you got a minute, there are a lot of changes I think it is because of some formatting rule changes? Not sure if I've done something wrong so please let me know if that is the case. The main changes happened in `Combobox` and `Listbox` components that can be found in the changes.

jesseleite commented:
Hoping this one can still be looked at sometime! 🙏 ... I haven't tested since conflicts were resolved the last time, but it definitely fixes our problem with [the command palette feature coming in Statamic](https://github.com/statamic/cms/pull/11699) ❤️ 

![image](https://github.com/user-attachments/assets/c2d013e7-c63b-4896-af20-bddc9afdde21)

zernonia commented:
Yo @J-Michalek ! can you perhaps resolves the conflicts? I'll review this soon 😁🙏🏼

zernonia committed:
fix: virtual list, cleanup

zernonia reviewed (approved):
Thanks for the PR guys! LGTM!

meirroth commented:
You guys just made my day!!! 🎉🚀

jesseleite commented:
Yeah, thank you so much for your work on this one!
